### PR TITLE
[MRG] Add function that returns all defined channel types

### PR DESCRIPTION
**Pull request by cbrnr**
The function `get_channel_types` returns a dict with known channel types as keys and their definitions in the `info` dict as values. This PR has two benefits:

1. It is now possible to ask for a list of known channel types
2. The function `channel_type` is now more generic and much shorter (because the logic is now in the dict defined in `get_channel_types`

WDYT? I can't believe I had to use the `else` clause of a `for` loop.

### CODE REVIEW

**Comment by cbrnr** [comment]
A use case for this functionality is described in https://github.com/cbrnr/mnelab/issues/5

**Comment by larsoner** [comment]
Looks reasonable to me, but [not travis](https://travis-ci.org/mne-tools/mne-python/jobs/318579085).

Can you check for relative speed improvements or regressions? I know the function overhead will be small but it might be used a lot of places in the code. Using `OrderedDict` with channels ordered roughly by how often they get picked (EEG, MEG, STIM, ECG, EOG?) entered first might make things faster in most use cases, for example.

Did you look at `channel_indices_by_type` to see if it would suit your use case already?

If you have time, can you look to see if there is a way to (finally!) deduplicate the logic here with that of `pick_types` now that you've done some standardization?

**Comment by cbrnr** [comment]
> Looks reasonable to me, but not travis.

Should be fixed. I assumed that the keys always existed.

> Can you check for relative speed improvements or regressions? I know the function overhead will be small but it might be used a lot of places in the code. Using `OrderedDict` with channels ordered roughly by how often they get picked (EEG, MEG, STIM, ECG, EOG?) entered first might make things faster in most use cases, for example.

Let's check how often this function is called before prematurely optimizing.

> Did you look at `channel_indices_by_type` to see if it would suit your use case already?

No, I really need to know *all* supported channel types, not just the ones in the current data set.

> If you have time, can you look to see if there is a way to (finally!) deduplicate the logic here with that of `pick_types` now that you've done some standardization?

OMG, I thought this would be a nice and quick PR 😝 . Yes, I'll take a look, but let's first assert that my change is usable in the first place before deduplicating this.

**Comment by larsoner** [comment]
> Let's check how often this function is called before prematurely optimizing.

Right now `git grep "channel_type("` gives 27 direct results. Functions that call those could multiply the effect. But it's not currently used by `pick_*` functions which will help limit its impact.

A quick check would be to try something like `pytest mne/io` before and after the PR to make sure nothing big changes.

When combined with the deduplication idea (something we should do eventually), it would be used any time a `pick_*` function would be used, though. So we'd need to look at speed then.

> OMG, I thought this would be a nice and quick PR

Yes that's why I added the "if you have time" -- it's not essential for this PR, but if you're in the mode to do it, it would be appreciated :)

**Comment by cbrnr** [comment]
All tests (except the OSX ones) pass. Now on to the timing question.

**Comment by cbrnr** [comment]
Master branch:
```
%timeit [mne.io.pick.channel_type(raw.info, idx) for idx in range(raw.info["nchan"])]
86.8 µs ± 1.51 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

This PR:
```
%timeit [mne.io.pick.channel_type(raw.info, idx) for idx in range(raw.info["nchan"])]
88.5 µs ± 2.11 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
`pytest mne/io` seems to be too variable to use as a reliable estimate of speed changes of this PR.

File: mne/io/pick.py [hunk 1]
```diff
@@ -31,54 +71,17 @@ def channel_type(info, idx):
...
-    elif kind == FIFF.FIFFV_FNIRS_CH:
-        if info['chs'][idx]['coil_type'] == FIFF.FIFFV_COIL_FNIRS_HBO:
-            return 'hbo'
-        elif info['chs'][idx]['coil_type'] == FIFF.FIFFV_COIL_FNIRS_HBR:
-            return 'hbr'
+    ch = info['chs'][idx]
+
+    for t, rules in get_channel_types().items():
+        for key, vals in rules.items():
+            if isinstance(vals, int) and ch.get(key, None) != vals:
```
Comment (agramfort) [inline mne/io/pick.py]: I find the logic hard to parse. Can you simplify the testing?

Comment (larsoner) [reply mne/io/pick.py]: for instance changing this to `if ch.get(key, None) not in np.array(vals)` might avoid the need for the conditional (depending on how well separated the logic currently is)

Comment (cbrnr) [reply mne/io/pick.py]: Yes, this gets rid of the condition. @larsoner I was wondering if this is possible with simple lists, by apparently there is no list constructor that does something similar to `np.array`, or is there?
@agramfort I've also added comments. Is the logic clearer now?

**Comment by cbrnr** [comment]
And yes, as @larsoner mentioned deduplicating this is important! I just saw that this is also used in `meas_info.py` in `create_info`. Let's not forget to to that after this PR.

**Comment by agramfort** [comment]
+1 for MRG when CIs are happy
